lib/min_heap.rb

```diff
@@ -44,28 +49,43 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1) 
+  # Space complexity: O(1) 
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return @store.empty?
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(log n)
+  # Space complexity: O(1)
   def heap_up(index)
-    
+    return if index == 0
+    until index == 0 || @store[index].key >= @store[((index - 1) / 2)].key
+      parent_index = (index - 1) / 2
+      swap(index, parent_index) if @store[parent_index].key > @store[index].key
+      index = parent_index
+    end
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    left = (2 * index) + 1
+    right = (2 * index) + 2
+    if !@store[left] && !@store[right]
+      return 
+    elsif !@store[right]
+      min = left
+    else
+      min = [left, right].min_by {|x| @store[x].key}
+    end
+    swap(index, min)
```

Comment:
This isn't quite working because you're not checking to see if the element at index is greater than the element at min.

```ruby
 if @store[index] > @store[min]
    swap(index, min)
    heap_down(min)
end
```